File: packaging/mcpb/swift/Sources/StackQLMCP/Platform.swift
import Foundation

/// Packaging platform keys. These match the stackql-mcpb-packaging release
/// asset names (`stackql-mcp-<key>.mcpb`) and the shared binary cache layout
/// used by the npm and pypi wrappers and the Go/Rust/Kotlin siblings. Both
/// macOS architectures map to the single universal binary.
public enum Platform: String, Sendable, CaseIterable {
    case linuxX64 = "linux-x64"
    case linuxArm64 = "linux-arm64"
    case windowsX64 = "windows-x64"
    case darwinUniversal = "darwin-universal"

    /// The platform key for an arbitrary OS/arch pair, or nil when no binary
    /// is published for it. Split out from `current` so it is unit-testable
    /// without depending on the host.
    public static func key(os: HostOS, arch: HostArch) -> Platform? {
        switch (os, arch) {
        case (.linux, .x86_64): return .linuxX64
        case (.linux, .arm64): return .linuxArm64
        case (.windows, .x86_64): return .windowsX64
        case (.macOS, .x86_64), (.macOS, .arm64): return .darwinUniversal
        default: return nil
        }
    }

    /// The platform key for the running host. nil on a platform with no
    /// published binary (for example Windows on arm64).
    public static var current: Platform? {
        key(os: HostOS.current, arch: HostArch.current)
    }

    /// The file name the extracted server binary uses in the shared cache.
    var executableName: String {
        self == .windowsX64 ? "stackql.exe" : "stackql"
    }
}

/// Host operating system, resolved from compile-time conditionals so it can
/// also be supplied explicitly in tests.
public enum HostOS: Sendable {
    case macOS
    case linux
    case windows
    case other

    public static var current: HostOS {
        #if os(macOS)
        return .macOS
        #elseif os(Linux)
        return .linux
        #elseif os(Windows)
        return .windows
        #else
        return .other
        #endif
    }
}

/// Host CPU architecture.
public enum HostArch: Sendable {
    case x86_64
    case arm64
    case other

    public static var current: HostArch {
        #if arch(x86_64)
        return .x86_64
        #elseif arch(arm64)
        return .arm64
        #else
        return .other
        #endif
    }
}

File: packaging/mcpb/cargo/src/platform.rs
use crate::error::{Error, Result};

/// A platform the packaging repo publishes a .mcpb bundle for.
///
/// Keys mirror stackql/stackql-mcpb-packaging: linux-x64, linux-arm64,
/// windows-x64, darwin-universal.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Platform {
    LinuxX64,
    LinuxArm64,
    WindowsX64,
    DarwinUniversal,
}

impl Platform {
    /// Detect the host platform, or fail if no bundle is published for it.
    pub fn detect() -> Result<Self> {
        Self::from_os_arch(std::env::consts::OS, std::env::consts::ARCH)
    }

    pub(crate) fn from_os_arch(os: &'static str, arch: &'static str) -> Result<Self> {
        match (os, arch) {
            ("linux", "x86_64") => Ok(Platform::LinuxX64),
            ("linux", "aarch64") => Ok(Platform::LinuxArm64),
            ("windows", "x86_64") => Ok(Platform::WindowsX64),
            // The darwin bundle is a universal binary, so any macOS arch works.
            ("macos", _) => Ok(Platform::DarwinUniversal),
            _ => Err(Error::UnsupportedPlatform { os, arch }),
        }
    }

    /// The platform key used in bundle names and cache paths.
    pub fn key(self) -> &'static str {
        match self {
            Platform::LinuxX64 => "linux-x64",
            Platform::LinuxArm64 => "linux-arm64",
            Platform::WindowsX64 => "windows-x64",
            Platform::DarwinUniversal => "darwin-universal",
        }
    }

    /// Name of the server binary inside the bundle's server/ directory.
    pub fn binary_name(self) -> &'static str {
        match self {
            Platform::WindowsX64 => "stackql.exe",
            _ => "stackql",
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn known_platforms_map_to_keys() {
        assert_eq!(
            Platform::from_os_arch("linux", "x86_64").unwrap().key(),
            "linux-x64"
        );
        assert_eq!(
            Platform::from_os_arch("linux", "aarch64").unwrap().key(),
            "linux-arm64"
        );
        assert_eq!(
            Platform::from_os_arch("windows", "x86_64").unwrap().key(),
            "windows-x64"
        );
        assert_eq!(
            Platform::from_os_arch("macos", "aarch64").unwrap().key(),
            "darwin-universal"
        );
        assert_eq!(
            Platform::from_os_arch("macos", "x86_64").unwrap().key(),
            "darwin-universal"
        );
    }

    #[test]
    fn unsupported_platform_is_an_error() {
        assert!(Platform::from_os_arch("freebsd", "x86_64").is_err());
        assert!(Platform::from_os_arch("windows", "aarch64").is_err());
    }

    #[test]
    fn windows_binary_has_exe_suffix() {
        assert_eq!(Platform::WindowsX64.binary_name(), "stackql.exe");
        assert_eq!(Platform::LinuxX64.binary_name(), "stackql");
    }
}
